compiler/mangler.nim

```diff
@@ -0,0 +1,704 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+# ------------------------- Name Mangling --------------------------------
+##[
+
+cgendata defines ConflictsTable as Table[string, int]
+
+The key has one of two forms:
+  "someMangledName" -> the output of mangle() on a symbol or
+  "someMangledName" -> the output of getTypeName() on a type or
+  "######" -> a series of digits representing a unique symbol or type id
+
+The values have different meanings depending upon the form of the key:
+  "someMangledName" -> the value represents the next available counter
+  "######" -> the value represents the counter associated with the id
+
+The counter is used to keep track of the order of symbol additions to the
+conflict table; they are increasing but not guaranteed to be sequential.
+
+]##
+
+import # compiler imports
+
+    ast, cgendata, modulegraphs, ropes, ccgutils, ndi, msgs,
+    idents, options, wordrecg, astalgo, treetab, sighashes
+
+import # stdlib imports
+
+  std / [ strutils, tables, sets ]
+
+const
+  inspect = "TGenericSeq"
+  debugMangle = false and not defined(release)
+  symbolicTypes =
+    when true:
+      {tyObject, tyEnum} # obey the boss
+    else:
+      {TTypeKind.low .. TTypeKind.high} - {tyString}
+
+  # these may be shorted to the first three characters following `ty`
+  # as opposed to perhaps being stripped of vowels, etc.
+  threeLetterShorties = {tyOrdinal, tySequence, tyObject, tyTuple, tyArray,
+                         tyVar, tyPtr, tyString, tySet, tyRef, tyTypeDesc}
+  # these are types that we ignore entirely when mangling
+  irrelevantForNaming = irrelevantForBackend + {tyVar} - {tyRange}
+  # these types may prefix the types they wrap, eg. `seqStr` or `setInt`
+  abbrevTypeArg = {tySet, tySequence, tyTypeDesc, tyArray, tyRange}
+  # these types get unwrapped and discarded from mangling
+  unwrapTypeArg = irrelevantForNaming - {tyRange} + {tyRef, tyPtr,
+                                         tyUserTypeClass, tyUserTypeClassInst}
+  unIdentity = ItemId(module: -1, item: -1)
+
+type
+  ModuleOrProc* = BProc or BModule
```

Comment:
This means all this stuff just generic with two different instantations. For no good reason. Please remove this and keep the old structure, generating a C type doesn't need a BProc context, so don't pretend that it does for your hacking conveniences.

Reply:
I didn't need this typeclass for 10 years. What do you need it for?

Reply:
I know, I didn't need it for 10 years, either.  Then I discovered this new mangling code and it came with functionality that is generic to `BProc` and `BModule`.  The way the author solved this is by creating a typeclass to allow the same mechanisms to update the `sigConflicts` structure that is duplicated in each of these types.

If you're happy with the backend from 10 years ago, we can go there.  But I'm personally annoyed that I have to play games because the proc signature is generated with the module instead of the proc, for example.  I would like to step away from the old code and embrace abstraction in the interests of evolution.  To me, it's worth a second copy of the proc in the nim compiler binary.

New code is new.  That's why we call it new.  It's due to the newness of the thing.

Reply:
New code can be better code. But this is so hacky that it's unacceptable. We all want better name mangling but I look at this Nim code that you wrote far more often than I look at the produced C code so it's a bad trade-off.

Reply:
Okay, well, if you come up with a better solution, I'll take another crack at it.